ElkM1 integration add types, part 5

gwww opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add stricter types to config_flow.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gwww committed:
Add config_flow to strict

gwww committed:
Add config_flow to strict

gwww committed:
ElkM1 integration, add types, part 5

gwww committed:
ElkM1 integration, add types, part 5

gwww commented on homeassistant/components/elkm1/config_flow.py:
```diff
@@ -279,7 +281,7 @@ async def async_step_discovered_connection(self, user_input=None):
                 user_input[CONF_PREFIX] = ""
             errors, result = await self._async_create_or_error(user_input, False)
-            if not errors:
+            if result is not None:
```
I checked all the code paths in `_async_create_or_error`, switching this check from error to result is cool, AFAICT.

gwww commented on homeassistant/components/elkm1/config_flow.py:
```diff
@@ -314,7 +318,7 @@ async def async_step_manual_connection(self, user_input=None):
                 user_input[CONF_ADDRESS] = device.ip_address
             errors, result = await self._async_create_or_error(user_input, False)
-            if not errors:
+            if result is not None:
```
Same as previous comment.

gwww commented on homeassistant/components/elkm1/config_flow.py:
```diff
@@ -357,7 +363,7 @@ async def async_step_import(self, user_input):
             return self.async_abort(reason=list(errors.values())[0])
         return result
```
Wasn't sure of the "right" fix here, without reworking `_async_create_or_error`. Typing wants me to check not None on each of the two return values. i.e.: to satisfy typing the code would have to look something like:
```
if not error:
    return error
if not result:
    return result
return ??? (must be a FlowResult)
```

epenet replied:
To help mypy understand these cases, you can add an assert statement:
```python
    if result is not None:
        return result
    assert errors is not None
    return self.async_abort(reason=list(errors.values())[0])
```

gwww committed:
Elk types 4 (#70305)

mib1185 committed:
Highlight to drag-and-drop diagnostic data instead of copy-and-paste in issue template (#70279)

PoltoS committed:
Scale Z-Wave.Me max values (#70274)

Co-authored-by: Dmitry Vlasov <kerbalspacema@gmail.com>

frenck committed:
Add backup.create service (#70118)

Flameeyes committed:
Don't attempt to set invalid fan percentage and preset_mode (try 3) (#70294)

Co-authored-by: J. Nick Koston <nick@koston.org>

justinedelson committed:
Add cmake to devcontainer (#70253)

* add cmake to devcontainer. fixes #70250

* reformatting

Kane610 committed:
Enable strict typing for deCONZ platforms (#70290)

fabaff committed:
Remove myself from bitcoin codeowners (#70315)

thecode committed:
Revert "Add humidity to tomorrowio hourly weather forecasts (#70186)" (#70311)

This reverts commit febcab11fbe43258f6553739c20f7232a662207d.

fabaff committed:
Remove myself from arest codeowners (#70314)

bdraco committed:
Bump aiohomekit to 0.7.17 (#70313)

Changelog: https://github.com/Jc2k/aiohomekit/compare/0.7.16...0.7.17

Fixes: #67665

frenck committed:
Replace Climate HVAC_MODE_* constants with HVACMode enum (#70286)

Co-authored-by: J. Nick Koston <nick@koston.org>

gwww committed:
Fix type error

tetienne committed:
Fix opening/closing for awning in Overkiz integration (#68890)

fabaff committed:
Upgrade psutil to 5.9.0 (#70316)

akloeckner committed:
Make `this` variable available in template entities (#65201)

* feat: make this variable available in template entities

This makes the variable `this` available in template entities.
It will simplify the use of self-referencing template entities.
Because, without this, we have to repeat the entity id every time.
If we can solve this without explicitly spelling the entity id,
code can be re-used much better.

As a side-effect, this will allow to use `variables`-like patterns,
where attributes can be used as variables to calculate subsequent attributes or state.

Example:
```yaml
template:
  sensor:
    - name: test
      state: "{{ this.attributes.test }}"
      # not: "{{ state_attr('sensor.test', 'test' }}"
      attributes:
        test: "{{ now() }}"
```

* expose entity_id instead of this

* add test

* Refactor to expose this variable

* Tweak repr dunder

Co-authored-by: Erik <erik@montnemery.com>

gwww committed:
Merge branch 'elk-types-5' of github.com:gwww/core into elk-types-5
